reader-privacy.md

```diff
@@ -0,0 +1,142 @@
+
+- [Introduction](#introduction)
+- [Background](#background)
+- [Specification](#specification)
+    - [Security](#security)
+- [Related Resources](#related-resources)
+​
+## Introduction
+​
+IPFS is currently lacking of many privacy protections. One of its main weak points lies in the lack 
+of privacy protections for the content routing subsystem. Currently neither readers (clients accessing files) 
+nor writers (hosts storing and distributing content) have much privacy with regard to content they publish or 
+consume. It is very easy for a content router node or a passive observer to learn which file is requested by 
+which client during the routing process, as the potential adversary easily learns about the requested `CID`. 
+A curious actor could request the same `CID` and download the associated file to monitor the user’s behavior. 
+This is obviously undesirable and has been for some time now a strong request from the community.
+
+The changes described in this specification introduce a IPNI Readres Privacy upgrade. It will prevent 
+passive observers from tracking user's actions as described above. It will also be a first step towards
+fully private IPNI protocol that will eliminate indexers as centralised observers.  
+
+### Non Goals
+
+* Writer, i.e. content provider or publisher, Privacy, which will be done in a separate specification
+* Retrieval Privacy, which is out of scope for the content routing subsystem.
+​
+## Background
+​
+Network indexers build their indexes by ingesting chains of Advertisements. Advertisement is a
+construct that allows Storage Providers to publish their CIDs in bulk (FIL deals) instead of doing
+that individually for each CID. A group of CIDs is represented by a unique ContextID as can be seen
+on the diagram below:
+​
+![Index building flow](resources/readers-privacy-1.png)
+
+## Specification
+​
+This specification improves the reader privacy by proposing changes to the Step 3, depicted above, where the client 
+supplies the content CID directly in order to lookup its corresponding providers.
+
+In order to protect the reader's privacy the proposal changes the way CID lookup works to the following:
+
+* A client who wants to do a lookup will calculate a hash over the CID's multihash (`hash(MH)`) and use it for the
+lookup query (hence the name double hashing);
+* In response to the hashed find request, the indexer will return a set of encrypted `ProviderRecordKey`s. 
+`ProviderRecordKey` will consist of two concatenated hashes - one over `peerID` and the other over `contextID`. 
+Each `ProviderRecordKey` will be encrypted with a key derived from the *original* multihash value: 
+`enc(hash(peerID) || hash(contextID), MH)`, where `hash` is a hash over the value, and `||` is concatenation 
+and `enc` is encryption over the value. In order to make sense of that payload, a passive observer would need 
+to get hold of the original CID that isn't revealed during the communication round;
```

Comment:
Instead of `enc(hash(peerID) || hash(contextID), MH)` can we use`enc(peerID || contextID, MH)` ?

then the client decodes `peerID` and calculates `hash(peerID)` for it's subsequent query

Reply:
The problem with full peer IDs in datastore keys (besides extra space required) is that they will make value keys be variable length. All current datastores rely on keys being fixed length so that they can be efficiently stored in arrays. 

Is your concern hash collisions?

Reply:
my concerns are:
1. Getting back the original peerID from the query means that this can continue to fullfill the current ContentRouting API in IPFS
2. Compatibility with the form that double-hashing for DHT is taking.

At one point masih proposed an additional level of 'pointer' indirection: in primary values table we point to another table of these `enc(ProviderRecordKey)` values - such that the record keys can be of variable length. (when used as keys, the key form is still `hash(ProviderRecordKey)`, so will remain fixed length)

In practice these should be almost all the same length - we could potentially partition the value table based on `len(ProviderRecordKey)` and have a few different value tables where we search each of them? (i imagine we'd have one for when a deal id is specified, and one for entries without a contextID))

Reply:
Added to the latest revision of the spec.